Sources/Logging/Logging.swift

```diff
@@ -665,28 +659,54 @@ public enum LoggingSystem {
     /// - parameters:
     ///     - factory: A closure that given a `Logger` identifier, produces an instance of the `LogHandler`.
     public static func bootstrap(_ factory: @escaping (String) -> LogHandler) {
-        #if canImport(WASILibc)
-        precondition(!self.initialized, "logging system can only be initialized once per process.")
-        self.factory = factory
-        self.initialized = true
-        #else
-        self.lock.withWriterLock {
-            precondition(!self.initialized, "logging system can only be initialized once per process.")
-            self.factory = factory
-            self.initialized = true
-        }
-        #endif
+        self._factory.replaceFactory(factory, validate: true)
     }
 
     // for our testing we want to allow multiple bootstraping
     internal static func bootstrapInternal(_ factory: @escaping (String) -> LogHandler) {
+        self._factory.replaceFactory(factory, validate: false)
+    }
+
+    fileprivate static var factory: (String) -> LogHandler {
+        return self._factory.underlying
+    }
+
+    private final class FactoryBox {
         #if canImport(WASILibc)
-        self.factory = factory
+        // WASILibc is single threaded, provides no locks
         #else
-        self.lock.withWriterLock {
-            self.factory = factory
-        }
+        private let lock = ReadWriteLock()
         #endif
+        fileprivate var _underlying: (String) -> LogHandler
+        private var initialized = false
+
+        init(_ underlying: @escaping (String) -> LogHandler) {
+            self._underlying = underlying
+        }
+
+        func replaceFactory(_ factory: @escaping (String) -> LogHandler, validate: Bool) {
+            #if canImport(WASILibc)
```

Comment:
shouldn't we push that into the ReaderWriterLock and just do nothing in WASI?

Reply:
l kept this "as is" based on the comment https://github.com/apple/swift-log/pull/222/files#diff-f5985f666e1e706d02b520068f87ee07c97e376122fbd685314e088e0d532bb0R676

Reply:
Yeah, but we can just put the do-nothing code into the `ReaderWriterLock`. That has less logic and is already `#if`-def'd for Windows etc already, right?

Reply:
i see, yes that is true